uniffi_bindgen/src/lib.rs

```diff
@@ -115,6 +115,133 @@ use bindings::TargetLanguage;
 use interface::ComponentInterface;
 use scaffolding::RustScaffolding;
 
+/// A trait representing a Binding Generator Configuration
+///
+/// External crates that implement binding generators will need to implement this trait
+/// with a type that implements [`serde::Deserialize`]. This trait represents what gets parsed from the
+/// uniffi.toml
+///
+/// # Example
+/// If the type that implements BindingGeneratorConfig is
+/// ```rs
+/// struct Config {
+///     gecko: GeckoConfig
+/// }
+/// ```
+/// Then, in the `uniffi.toml` the inner `gecko` is accessible using the toml key `gecko`, for example:
+/// ```toml
+/// [gecko]
+/// ```
+pub trait BindingGeneratorConfig:
+    for<'de> Deserialize<'de>
+    + MergeWith
+    + for<'a> From<&'a ComponentInterface>
+    + Default
+    + std::fmt::Debug
+{
+    fn load<P: AsRef<Path>>(
+        ci: &ComponentInterface,
+        crate_root: P,
+        config_file_override: Option<P>,
+    ) -> anyhow::Result<Self> {
+        let config: Self = ci.into();
+        let config_file: Option<PathBuf> = match config_file_override {
+            Some(cfg) => Some(PathBuf::from(cfg.as_ref())),
+            None => crate_root.as_ref().join("uniffi.toml").canonicalize().ok(),
+        };
+        match config_file {
+            Some(path) => {
+                let contents = slurp_file(&path)
+                    .with_context(|| format!("Failed to read config file from {:?}", &path))?;
+                let loaded_config: Self = toml::de::from_str(&contents)
+                    .with_context(|| format!("Failed to generate config from file {:?}", &path))?;
```

Comment:
I think there are a couple of big drawbacks for having the bindings code see the entire `uniffi.toml` config:

  - The `From<&ComponentInterface>` impl needs to return it.  This means after it does the real work, it also needs to build a `bindings::Config`, but with empty values for the other bindings languages.  Then put that into a `crate::Config`.
  - The `merge_with` code needs to merge the entire config, rather than just the bindings-specific config.
  
It doesn't seem good to force bindings authors to write all that code out.

Reply:
I see, just to clarify I understand (I'm new to how the configs are handled, so feel free to shut this down!):
It's a big issue because we would need to build the entire Config, and fill in dummy values for everything we don't know about (i.e other config keys, etc)?

What I'm thinking is that external crates would **not** see any global config, instead, each bindings have their own config. We can have another one (if we need it?) that lives privately in `uniffi`, which represents the keys that uniffi cares about that are not specific to any binding.

The idea is that `From<&ComponentInterface>` (or a constructor 🤷) doesn't build a `crate::Config`, instead it builds the type that implements `BindingGeneratorConfig` directly, and the same with `merge_with`, it would only merge the binding-specific configs

I understand that there is a bit of complexity here that might not be worth it, and I might be missing something else so I'm happy to change it up as well!

Reply:
> The idea is that From<&ComponentInterface> (or a constructor shrug) doesn't build a uniffi_bindgen::Config, instead it builds the type that implements BindingGeneratorConfig directly, and the same with merge_with, it would only merge the binding-specific configs

I think this part would work fine (although I feels like there's a cleaner solution that doesn't use the `From` and `MergeWith` traits).

But then how do you got from a path to a `.toml` file to the bindings-specific config?  Right now they way it works, is that we embed the bindings-specific config as a field and let serde do everything.  But once crates are external that becomes difficult.  I think we need a method on either `BindingsGenerator` or `BindingsGeneratorConfig` that does this.  I'll try to push some code to make it more clear.